src/edu/pugetsound/mathcs/architecture_nlp/features/spf/TrainModel.java

```diff
@@ -0,0 +1,430 @@
+import edu.cornell.cs.nlp.spf.parser.ccg.cky.CKYBinaryParsingRule;
+import edu.cornell.cs.nlp.spf.parser.ccg.cky.CKYUnaryParsingRule;
+import edu.cornell.cs.nlp.spf.parser.ccg.cky.logger.ChartLogger;
+import edu.cornell.cs.nlp.spf.parser.ccg.cky.single.CKYParser;
+import edu.cornell.cs.nlp.spf.parser.ccg.cky.sloppy.BackwardSkippingRule;
+import edu.cornell.cs.nlp.spf.parser.ccg.cky.sloppy.ForwardSkippingRule;
+import edu.cornell.cs.nlp.spf.parser.ccg.cky.sloppy.SimpleWordSkippingLexicalGenerator;
+import edu.cornell.cs.nlp.spf.parser.ccg.factoredlex.features.FactoredLexicalFeatureSet;
+import edu.cornell.cs.nlp.spf.parser.ccg.features.basic.DynamicWordSkippingFeatures;
+import edu.cornell.cs.nlp.spf.parser.ccg.features.basic.LexicalFeaturesInit;
+import edu.cornell.cs.nlp.spf.parser.ccg.features.basic.scorer.ExpLengthLexicalEntryScorer;
+import edu.cornell.cs.nlp.spf.parser.ccg.features.lambda.LogicalExpressionCoordinationFeatureSet;
+import edu.cornell.cs.nlp.spf.parser.ccg.lambda.pruning.SupervisedFilterFactory;
+import edu.cornell.cs.nlp.spf.parser.ccg.model.LexiconModelInit;
+import edu.cornell.cs.nlp.spf.parser.ccg.model.Model;
+import edu.cornell.cs.nlp.spf.parser.ccg.model.ModelLogger;
+import edu.cornell.cs.nlp.spf.parser.ccg.normalform.NormalFormValidator;
+import edu.cornell.cs.nlp.spf.parser.ccg.normalform.hb.HBComposedConstraint;
+import edu.cornell.cs.nlp.spf.parser.ccg.rules.lambda.PluralExistentialTypeShifting;
+import edu.cornell.cs.nlp.spf.parser.ccg.rules.lambda.ThatlessRelative;
+import edu.cornell.cs.nlp.spf.parser.ccg.rules.lambda.typeraising.ForwardTypeRaisedComposition;
+import edu.cornell.cs.nlp.spf.parser.ccg.rules.lambda.typeshifting.PrepositionTypeShifting;
+import edu.cornell.cs.nlp.spf.parser.ccg.rules.primitivebinary.application.BackwardApplication;
+import edu.cornell.cs.nlp.spf.parser.ccg.rules.primitivebinary.application.ForwardApplication;
+import edu.cornell.cs.nlp.spf.parser.ccg.rules.primitivebinary.composition.BackwardComposition;
+import edu.cornell.cs.nlp.spf.parser.ccg.rules.primitivebinary.composition.ForwardComposition;
+import edu.cornell.cs.nlp.spf.parser.graph.IGraphParser;
+import edu.cornell.cs.nlp.spf.test.Tester;
+import edu.cornell.cs.nlp.spf.test.stats.ExactMatchTestingStatistics;
+import edu.cornell.cs.nlp.utils.collections.SetUtils;
+import edu.cornell.cs.nlp.utils.function.PredicateUtils;
+import edu.cornell.cs.nlp.utils.log.ILogger;
+import edu.cornell.cs.nlp.utils.log.Log;
+import edu.cornell.cs.nlp.utils.log.LogLevel;
+import edu.cornell.cs.nlp.utils.log.Logger;
+import edu.cornell.cs.nlp.utils.log.LoggerFactory;
+/**
+ * Class to train a SPFModel. Using the data files in preset directories, this will train a new Sentence
+ * to LogicalExpression parser, and write the results to a file that can later be used by GenerateInteractor
+ * to incorporate the SPFModel into the current NLP pipeline
+ * 
+ * @author Yoav Artzi & Jared Polonitza
```

Comment:
Again, just add a comment explaining how this file differs from the corresponding class in the SPF jar file